src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -44,131 +42,50 @@ internal abstract partial class AbstractInheritanceMarginService : IInheritanceM
-                    derivedTypesSymbols: derivedSymbols.CastArray<ISymbol>(),
-                    cancellationToken).ConfigureAwait(false);
-                builder.AddIfNotNull(item);
-            }
+            return await serializedInheritanceMarginItems.SelectAsArrayAsync(
+                (serializedItem, _) => InheritanceMarginItem.ConvertAsync(solution, serializedItem, cancellationToken), cancellationToken).ConfigureAwait(false);
         }
 
-        private async Task AddInheritanceMemberItemsForTypeMembersAsync(
-            Solution solution,
-            TextLineCollection lines,
-            SyntaxNode declarationNode,
-            ISymbol memberSymbol,
-            ArrayBuilder<InheritanceMarginItem> builder,
-            CancellationToken cancellationToken)
-        {
-            // For a given member symbol (method, property and event), its base and derived symbols are classified into 4 cases.
-            // The mapping between images
-            // Implemented : I↓
-            // Implementing : I↑
-            // Overridden: O↓
-            // Overriding: O↑
-
-            // Go down the inheritance chain to find all the overrides targets.
-            var overriddenSymbols = await SymbolFinder.FindOverridesArrayAsync(memberSymbol, solution, cancellationToken: cancellationToken).ConfigureAwait(false);
-
-            // Go up the inheritance chain to find all the implemented targets.
-            var implementingSymbols = memberSymbol.ExplicitOrImplicitInterfaceImplementations();
-
-            // Go up the inheritance chain to find all overriding targets
-            var overridingSymbols = GetOverridingSymbols(memberSymbol);
-
-            // Go down the inheritance chain to find all the implementing targets.
-            var implementedSymbols = await GetImplementedSymbolsAsync(solution, memberSymbol, cancellationToken).ConfigureAwait(false);
-
-            if (overriddenSymbols.Any() || overridingSymbols.Any() || implementingSymbols.Any() || implementedSymbols.Any())
-            {
-                var identifierToken = GetDeclarationToken(declarationNode);
-                var lineNumber = lines.GetLineFromPosition(identifierToken.SpanStart).LineNumber;
-                var item = await CreateInheritanceMemberInfoForMemberAsync(
-                    solution,
-                    memberSymbol,
-                    lineNumber,
-                    implementingMembers: implementingSymbols,
-                    implementedMembers: implementedSymbols,
-                    overridenMembers: overriddenSymbols,
-                    overridingMembers: overridingSymbols,
-                    cancellationToken).ConfigureAwait(false);
-                builder.AddIfNotNull(item);
-            }
-        }
+        private static bool CanHaveInheritanceTarget(ISymbol symbol)
+            => !symbol.IsStatic && (symbol is INamedTypeSymbol or IEventSymbol or IPropertySymbol || symbol.IsOrdinaryMethod());
```

Comment:
fwiw, i find this sort of thing pretty confusing to read.  when there are lots of `&&` and `or` or `||`, then i prefer statements.  but it's up to you.